core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -78,6 +89,78 @@ public void prepareRun(BatchJoinerContext context) throws Exception {
+   * For each of these transformations, if the input field is directly coming from the schema
+   * of one of the stage, the field is added as {@code stage_name.field_name}. We keep track of fields
+   * outputted by operation (in {@code outputsSoFar set}, so that any operation uses that field as
+   * input later, we add it without the stage name.
+   *
+   * Join transform operation is added with join keys as input tagged with the stage name, and join keys
+   * without stage name as output.
+   *
+   * For other fields which are not renamed in join, Identity transform is added, while for fields which
+   * are renamed Rename transform is added.
+   **
+   * @param outputFieldInfos collection of output fields along with information such as stage name, alias
+   * @param perStageJoinKeys join keys
+   * @return List of field operations
+   */
+  @VisibleForTesting
+  public static List<FieldOperation> createFieldOperations(Collection<OutputFieldInfo> outputFieldInfos,
+                                                           Map<String, List<String>> perStageJoinKeys) {
+    LinkedList<FieldOperation> operations = new LinkedList<>();
+
+    // Add JOIN operation
+    List<String> joinInputs = new ArrayList<>();
+    Set<String> joinOutputs = new LinkedHashSet<>();
+    for (Map.Entry<String, List<String>> joinKey : perStageJoinKeys.entrySet()) {
+      for (String field : joinKey.getValue()) {
+        joinInputs.add(joinKey.getKey() + "." + field);
+        joinOutputs.add(field);
+      }
+    }
+    FieldOperation joinOperation = new FieldTransformOperation("Join", JOIN_OPERATION_DESCRIPTIOPN, joinInputs,
+                                                               new ArrayList<>(joinOutputs));
+    operations.add(joinOperation);
+
+    Set<String> outputsSoFar = new HashSet<>(joinOutputs);
+
+    for (OutputFieldInfo outputFieldInfo : outputFieldInfos) {
+      // input field name for the operation will come in from schema if its not outputted so far
+      String inputFieldName = outputsSoFar.contains(outputFieldInfo.inputFieldName) ?
+        outputFieldInfo.inputFieldName : outputFieldInfo.stageName + "." + outputFieldInfo.inputFieldName;
+
+      if (outputFieldInfo.name.equals(outputFieldInfo.inputFieldName)) {
+        // Record identity transform
+        if (perStageJoinKeys.get(outputFieldInfo.stageName).contains(outputFieldInfo.inputFieldName)) {
+          // if the field is part of join key no need to emit the identity transform as it is already taken care
+          // by join
+          continue;
+        }
+        String operationName = String.format("Identity %s", inputFieldName);
+        FieldOperation identity = new FieldTransformOperation(operationName, IDENTITY_OPERATION_DESCRIPTION,
+                                                              Collections.singletonList(inputFieldName),
+                                                              outputFieldInfo.inputFieldName);
```

Comment:
nit: I know they are the same value, but it is easier for me to read this as outputFieldInfo.name. Using 'inputFieldName' as the operation 'output' made me think a little bit.

Reply:
fixed.